58b390f21b triggers new build-time warning on threaded builds on FreeBSD

A recent commit has introduced a new build-time warning when running `make` on a threaded build on FreeBSD-12 using either `gcc10` or `clang10` as the compiler.

**Summary**

```
$ parse-one-warnings-group  e616482ebb.freebsd.threaded.clang.maketp.output.txt.gz 'Wunused-function'
File:  e616482ebb.freebsd.threaded.clang.maketp.output.txt.gz

[
  {
    char   => 1,
    group  => "Wunused-function",
    line   => 652,
    source => "locale.c",
    text   => "unused function 'S_less_dicey_bool_setlocale_r'",
  },
]
```

**Context**

```
clang -c -DPERL_CORE -DHAS_FPSETMASK -DHAS_FLOATINGPOINT_H -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -std=c99 -O2 -Wall -Werror=pointer-arith -Werror=vla -Wextra -Wno-long-long -Wno-declaration-after-statement -Wc++-compat -Wwrite-strings locale.c
locale.c:652:1: warning: unused function 'S_less_dicey_bool_setlocale_r' [-Wunused-function]
S_less_dicey_bool_setlocale_r(pTHX_ const int cat, const char * locale)
^
1 warning generated.
```
**Analysis**

I haven't bisected this but the line in question in `locale.c` was touched in this commit:
```
commit 58b390f21b8cbcf54ebe4dfb3786c00da473edd5
Author:     Karl Williamson <khw@cpan.org>
AuthorDate: Sun Mar 14 05:29:23 2021 -0600
Commit:     Karl Williamson <khw@cpan.org>
CommitDate: Wed Sep 21 14:34:41 2022 -0600

    locale.c: Mitigate unsafe threaded locales
```
This is *not* appearing in threaded builds on Debian 11 Linux using `gcc-10` as the compiler, but I don't know why that is so.

@khwilliamson, can you take a look?

Thank you very much.
Jim Keenan

--- a/embed.fnc
+++ b/embed.fnc
@@ -3389,9 +3389,12 @@ S	|const char *|find_locale_from_environment|const unsigned int index
 S	|const char *|less_dicey_setlocale_r				\
 				|const int category			\
 				|NULLOK const char * locale
+: Not currently used
+#      if 0
 S	|bool	|less_dicey_bool_setlocale_r				\
 				|const int cat				\
 				|NN const char * locale
+#      endif
 S	|void	|less_dicey_void_setlocale_i				\
 				|const unsigned cat_index		\
 				|NN const char * locale			\

--- a/embed.h
+++ b/embed.h
@@ -1528,10 +1528,18 @@
 #    endif
 #  endif
 #  if !(defined(USE_POSIX_2008_LOCALE))
+#    if 0
+#      if defined(PERL_IN_LOCALE_C)
+#        if defined(USE_LOCALE)
+#          if defined(USE_LOCALE_THREADS)					       && ! defined(USE_THREAD_SAFE_LOCALE)				       && ! defined(USE_THREAD_SAFE_LOCALE_EMULATION)
+#define less_dicey_bool_setlocale_r(a,b)	S_less_dicey_bool_setlocale_r(aTHX_ a,b)
+#          endif
+#        endif
+#      endif
+#    endif
 #    if defined(PERL_IN_LOCALE_C)
 #      if defined(USE_LOCALE)
 #        if defined(USE_LOCALE_THREADS)					       && ! defined(USE_THREAD_SAFE_LOCALE)				       && ! defined(USE_THREAD_SAFE_LOCALE_EMULATION)
-#define less_dicey_bool_setlocale_r(a,b)	S_less_dicey_bool_setlocale_r(aTHX_ a,b)
 #define less_dicey_setlocale_r(a,b)	S_less_dicey_setlocale_r(aTHX_ a,b)
 #define less_dicey_void_setlocale_i(a,b,c)	S_less_dicey_void_setlocale_i(aTHX_ a,b,c)
 #        endif

--- a/locale.c
+++ b/locale.c
@@ -646,6 +646,8 @@ S_less_dicey_void_setlocale_i(pTHX_ const unsigned cat_index,
 #  define void_setlocale_r(cat, locale)                                     \
        void_setlocale_i(get_category_index(cat, locale), locale)
 
+#  if 0     /* Not currently used */
+
 STATIC bool
 S_less_dicey_bool_setlocale_r(pTHX_ const int cat, const char * locale)
 {
@@ -660,6 +662,7 @@ S_less_dicey_bool_setlocale_r(pTHX_ const int cat, const char * locale)
     return retval;
 }
 
+#  endif
 #  define bool_setlocale_r(cat, locale)                                 \
                                less_dicey_bool_setlocale_r(cat, locale)
 #  define bool_setlocale_i(i, locale)                                   \

--- a/proto.h
+++ b/proto.h
@@ -4722,12 +4722,20 @@ STATIC const char *	S_calculate_LC_ALL(pTHX_ const char ** individ_locales);
 #  endif
 #endif
 #if !(defined(USE_POSIX_2008_LOCALE))
-#  if defined(PERL_IN_LOCALE_C)
-#    if defined(USE_LOCALE)
-#      if defined(USE_LOCALE_THREADS)					       && ! defined(USE_THREAD_SAFE_LOCALE)				       && ! defined(USE_THREAD_SAFE_LOCALE_EMULATION)
+#  if 0
+#    if defined(PERL_IN_LOCALE_C)
+#      if defined(USE_LOCALE)
+#        if defined(USE_LOCALE_THREADS)					       && ! defined(USE_THREAD_SAFE_LOCALE)				       && ! defined(USE_THREAD_SAFE_LOCALE_EMULATION)
 STATIC bool	S_less_dicey_bool_setlocale_r(pTHX_ const int cat, const char * locale);
 #define PERL_ARGS_ASSERT_LESS_DICEY_BOOL_SETLOCALE_R	\
 	assert(locale)
+#        endif
+#      endif
+#    endif
+#  endif
+#  if defined(PERL_IN_LOCALE_C)
+#    if defined(USE_LOCALE)
+#      if defined(USE_LOCALE_THREADS)					       && ! defined(USE_THREAD_SAFE_LOCALE)				       && ! defined(USE_THREAD_SAFE_LOCALE_EMULATION)
 STATIC const char *	S_less_dicey_setlocale_r(pTHX_ const int category, const char * locale);
 #define PERL_ARGS_ASSERT_LESS_DICEY_SETLOCALE_R
 STATIC void	S_less_dicey_void_setlocale_i(pTHX_ const unsigned cat_index, const char * locale, const line_t line);

